FileInstaller/FileInstaller.cpp

```diff
@@ -0,0 +1,106 @@
+﻿#include <windows.h>
+#include <iostream>
+#include <vector>
+
+#include "FileInstaller.hpp"
+#include "FileUtils.hpp"
+
+FileInstaller::FileInstaller(std::vector<std::wstring> &file_paths, std::wstring const &installation_dir)
+	: m_file_paths(file_paths),
+	m_installation_dir(installation_dir),
+	m_is_dir_already_exists(false),
+	m_file_paths_to_clean({}) {}
+
+FileInstaller::~FileInstaller() {}
+
+void FileInstaller::install() {
+	try {
+		this->_create_installation_dir();
+		this->_copy_files();
+	}
+	catch (FileInstallerException &e) {
+		this->_revert_installation();
+		// Installation exception will be ignored in case an exception will be thrown during revertion.
+		throw e;
+	}
+}
+
+void FileInstaller::_revert_installation() {
+	DEBUG_MSG("Reverting installation (best effort)");
+	// Because we're doing best effort we're not return any return values about the status of the revert operation.
+	(void)this->_delete_installed_files();
+
+	if (!this->m_is_dir_already_exists) {
+		(void)this->_delete_installation_dir();
+	}
+}
+
+void FileInstaller::_copy_file(std::wstring const &file_path) {
+	std::wstring full_target_path = FileUtils::change_file_path_directory(file_path, this->m_installation_dir);
+
+	bool is_file_exists = FileUtils::is_path_exists(full_target_path);
+	if (!is_file_exists) {
+		FileUtils::copy_file(file_path, full_target_path);
+		this->m_file_paths_to_clean.push_back(file_path.c_str());
```

Comment:
```push_back``` allocates memory and can throw exceptions on failure, 
if this happens your code will forget the file you just copied

Reply:
Replaced between the two lines so the push_back will before the copy.
Now the delete function when reverting will not crash if the file isn't exist when delete.